src/avro_decoder_hooks.erl

```diff
@@ -0,0 +1,272 @@
+%%      end
+%% end.
+%% </pre>
+%%
+%% A hook can be used for debug. For example, below hook should print
+%% the decoding stack along the decode function traverses through the bytes.
+%%
+%% <pre>
+%% fun(Type, SubNameOrIndex, Data, DecodeFun) ->
+%%      SubInfo = case is_integer(SubNameOrIndex) of
+%%                  true  -> integer_to_list(SubNameOrIndex);
+%%                  false -> SubNameOrIndex
+%%                end,
+%%      io:format("~s.~s\n", [avro:get_type_name(Type), SubInfo]),
+%%      DecodeFun(Data)
+%% end
+%% </pre>
+%%
+%% A hook can also be used as a monkey patch to fix some corrupted data.
+%%% @end
+%%%-------------------------------------------------------------------
+-module(avro_decoder_hooks).
+
+-export([ tag_unions/0
+        , pretty_print_hist/0
+        , print_debug_trace/2
+        ]).
+
+-include("avro_internal.hrl").
+
+-define(PD_PP_INDENTATION, '$avro_decoder_pp_indentation').
+-define(PD_DECODER_HIST, '$avro_decoder_hist').
+-define(REASON_TAG, '$hook-raised').
+
+-type count() :: non_neg_integer().
+-type trace_hist_entry() :: {push, _, _} | {pop, _} | pop.
+
+%% @doc By default, decoders do not tag union values.
+%% This hook function is to tag union values with union type names
+%% NOTE: null values are not tagged
+%% @end
+-spec tag_unions() -> decoder_hook_fun().
+tag_unions() -> fun tag_unions/4.
+
+%% @doc This hook is useful when a decoder has failed on decoding,
+%% try to decode it again with this hook to inspect the decode history
+%% and the avro type stack where the failure happened
+%% NOTE: Always call this API to retrieve the hook, never save the hook
+%%       and re-use for different decode attempts
+%% @end.
+-spec print_debug_trace(fun((iodata()) -> ok), count()) -> decoder_hook_fun().
+print_debug_trace(PrintFun, MaxHistoryLength) ->
+  ok = erase_hist(),
+  fun(T, Sub, Data, DecodeFun) ->
+    print_trace_on_failure(T, Sub, Data, DecodeFun, PrintFun, MaxHistoryLength)
+  end.
+
+%% @doc Return a function to be used as the decoder hook.
+%% The hook prints the type tree with indentation, and the leaf values
+%% to the io-device 'user' is directed.
+%% @end
+-spec pretty_print_hist() -> decoder_hook_fun().
+pretty_print_hist() ->
+  _ = erase(?PD_PP_INDENTATION),
+  fun(T, SubInfo, Data, DecodeFun) ->
+    Name = avro:get_type_fullname(T),
+    Indentation =
+      case get(?PD_PP_INDENTATION) of
+        undefined -> 0;
+        Indentati -> Indentati
+      end,
+    IndentationStr = lists:duplicate(Indentation * 2, $\s),
+    ToPrint =
+      [ IndentationStr
+      , Name
+      , case SubInfo of
+          ""                   -> ": ";
+          I when is_integer(I) -> [$., integer_to_list(I), "\n"];
+          S when is_list(S)    -> [$., S, "\n"];
+          B when is_binary(B)  -> [$., B, "\n"];
+          _                    -> "\n"
+        end
+      ],
+    io:format(user, "~s", [ToPrint]),
+    _ = put(?PD_PP_INDENTATION, Indentation + 1),
```

Comment:
this hook was originaly for fun.
it is kept here mostly to demo how a hook traverses through the type tree.

The next one (`print_trace_on_failure`) is for serious use case, 
However it also uses pd for the same reason - simplicity.

We have another version in `bixporter` in which the traces are collected in another process,
but that one has to spawn a process first, (also needs some clean up work).
while using pd should allow the users to use the hook with only one call.

A 'nicer' alternative is to add an extra argument (e.g. `HookAcc`) and make all the
decode functions carry it recursively, also pass it into the `HookFun` and get a `NewHookAcc` for each `HookFun` return.
However, IMHO, it makes the decoding code even more complex, also makes it harder to write a hook.

I believe most of the use cases (such as the ones listed in README) for hooks do not require this HookAcc.